pkg/upload/upload.go

```diff
@@ -0,0 +1,280 @@
+	reference, err := file.SplitWriteAll(ctx, sp, bytes.NewReader(fileEntryBytes), int64(len(fileEntryBytes)), fileInfo.ToEncrypt)
+	if err != nil {
+		return swarm.ZeroAddress, fmt.Errorf("split entry error: %v", err)
+	}
+
+	return reference, nil
+}
+
+// DirUploadInfo contains the data for a dir to be uploaded
+type DirUploadInfo struct {
+	DefaultPath string
+	DirReader   io.ReadCloser
+}
+
+// GetDirHTTPInfo extracts dir info for upload from HTTP request
+// based on ethersphere/swarm/api/api.go/UploadTar
+func GetDirHTTPInfo(r *http.Request) (*DirUploadInfo, error) {
+	defaultPath := r.URL.Query().Get("defaultpath")
+	return &DirUploadInfo{
+		DefaultPath: defaultPath,
+		DirReader:   r.Body,
+	}, nil
+}
+
+// StoreDir stores all files contained in the given directory as a tar and returns its reference
+// based on ethersphere/swarm/api/api.go/UploadTar
+func StoreDir(ctx context.Context, dirInfo *DirUploadInfo, s storage.Storer) (swarm.Address, error) {
+	var contentKey swarm.Address // how is this determined?
+	// manifestPath := GetURI(r.Context()).Path // ??
+
+	bodyReader := dirInfo.DirReader
+	tr := tar.NewReader(bodyReader)
+	defer bodyReader.Close()
+
+	var defaultPathFound bool
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return swarm.ZeroAddress, fmt.Errorf("read tar stream error: %v", err)
+		}
+
+		// only store regular files
+		if !hdr.FileInfo().Mode().IsRegular() { // ??
+			continue
+		}
+
+		// manifestPath := path.Join(manifestPath, hdr.Name)
+		// apparently, `h.Xattrs[key] = value` == `h.PAXRecords["SCHILY.xattr."+key] = value`
+		contentType := hdr.PAXRecords["SCHILY.xattr."+"user.swarm.content-type"]
+		if contentType == "" {
+			contentType = mime.TypeByExtension(filepath.Ext(hdr.Name))
+		}
+
+		// add the entry under the path from the request
+		/*
+			entry := &ManifestEntry{
+				Path:        manifestPath,
+				ContentType: contentType,
+				Mode:        hdr.Mode,
+				Size:        hdr.Size,
+				ModTime:     hdr.ModTime,
+			}
+			contentKey, err = mw.AddEntry(ctx, tr, entry)
+			if err != nil {
+				return nil, fmt.Errorf("error adding manifest entry from tar stream: %s", err)
+			}*/
+
+		if hdr.Name == dirInfo.DefaultPath {
+			/*contentType := hdr.Xattrs["user.swarm.content-type"]
+			if contentType == "" {
+				contentType = mime.TypeByExtension(filepath.Ext(hdr.Name))
+			}
+
+			entry := &ManifestEntry{
+				Hash:        contentKey.Hex(),
+				Path:        "", // default entry
+				ContentType: contentType,
+				Mode:        hdr.Mode,
+				Size:        hdr.Size,
+				ModTime:     hdr.ModTime,
+			}
+			conctx context.Context,return nil, fmt.Errorf("error adding default manifest entry from tar stream: %s", err)
+			}*/
+
+			defaultPathFound = true
+		}
+
+		fileInfo := &FileUploadInfo{
+			FileName:    hdr.Name, // or hdr.FileInfo().Name() ??
+			FileSize:    hdr.Size,
+			ContentType: contentType,
+			ToEncrypt:   false,      // when could it be true?
```

Comment:
I guess it should be propagated from the API, if it should be encrypted or not.

Reply:
i ended up mimicking the `fileUploadHandler` behavior, and just requested a specific header on the dir upload HTTP request.

this is then passed down to all individual upload calls for every file contained in the tar file for the uploaded directory.